File: 19. CloneGraph.cpp
/*
Task: Clone Graph
Given a reference to a node in a connected undirected graph, return a deep copy (clone) of the
graph.
*/
#include <iostream>
#include <vector>
#include <unordered_map>
using namespace std;

struct Node
{
    int val;
    vector<Node*> neighbors;
    Node(int _val) : val(_val) {}
};
Node* cloneGraph(Node* node, unordered_map<Node*, Node*>& mp)
{
    if (!node)
        return nullptr;
    if (mp.count(node))
        return mp[node];
    Node* clone = new Node(node->val);
    mp[node] = clone;
    for (auto n : node->neighbors) clone->neighbors.push_back(cloneGraph(n, mp));
    return clone;
}
Node* cloneGraph(Node* node)
{
    unordered_map<Node*, Node*> mp;
    return cloneGraph(node, mp);
}
// Sample Test Case
int main()
{
    Node* n1 = new Node(1);
    Node* n2 = new Node(2);
    Node* n3 = new Node(3);
    Node* n4 = new Node(4);
    n1->neighbors = {n2, n4};
    n2->neighbors = {n1, n3};
    n3->neighbors = {n2, n4};
    n4->neighbors = {n1, n3};
    Node* clone = cloneGraph(n1);
    cout << "Cloned Node Val: " << clone->val << endl;
    return 0;
}
/*
Time Complexity: O(V+E)
Space Complexity: O(V)
*/

File: 19. CloneGraph.py
"""
Task: Clone Graph
Given a reference to a node in a connected undirected graph, return a deep copy (clone) of the graph.
"""
class Node:
    def __init__(self, val):
        self.val = val
        self.neighbors = []
def clone_graph(node, mp=None):
    if not node:
        return None
    if mp is None:
        mp = {}
    if node in mp:
        return mp[node]
    clone = Node(node.val)
    mp[node] = clone
    for n in node.neighbors:
        clone.neighbors.append(clone_graph(n, mp))
    return clone
# Sample Test Case
if __name__ == "__main__":
    n1 = Node(1)
    n2 = Node(2)
    n3 = Node(3)
    n4 = Node(4)
    n1.neighbors = [n2, n4]
    n2.neighbors = [n1, n3]
    n3.neighbors = [n2, n4]
    n4.neighbors = [n1, n3]
    clone = clone_graph(n1)
    print("Cloned Node Val:", clone.val)
"""
Time Complexity: O(V+E)
Space Complexity: O(V)
"""
